awscli/customizations/s3/s3handler.py

```diff
@@ -523,6 +525,80 @@ def _process_transfer(self, future):
             return CommandResult(1, 0)
 
 
+class S3TransferHandler(object):
+    def __init__(self, transfer_manager, cli_params, result_queue,
+                 result_recorder, result_processor):
+        """Backend for performing S3 transfers
+
+        :type transfer_manager: s3transfer.manager.TransferManager
+        :param transfer_manager: Transfer manager to use for transfers
+
+        :type cli_params: dict
+        :param cli_params: The parameters passed to the CLI command in the
+            form of a dictionary
+
+        :type result_queue: queue.Queue
+        :param result_queue: The queue to process results
+
+        :type result_recorder: ResultRecorder
+        :param result_recorder: Result recorder to get status of transfer
+
+        :type result_processor: ResultProcessor
+        :param result_processor: Result processor to process status updates
+            of transfer
+        """
+        self._transfer_manager = transfer_manager
+        self._result_queue = result_queue
```

Comment:
Seems like there's a missing abstraction here.  You have a result queue, recorder, and processor, but at this level of abstraction, it seems too low level.  Especially in methods like `_shutdown()` where you have to coordinate how to shutdown the "result subsystem" by first enqueueing a shutdown request, then joining on a thread.  Seems like this can be packaged behind a class.